configuration.rst

```diff
@@ -186,111 +215,353 @@ This parameter is then referenced in the framework config in
+* In the **test** environment, the ``config/packages/test/framework.yaml`` file
+  is loaded to override some of the settings previously configured in
+  ``config/packages/framework.yaml``.
+
+.. seealso::
+
+    See the ``configureContainer()`` method of
+    :doc:`the Kernel class </configuration/front_controllers_and_kernel>` to
+    learn everything about the loading order of configuration files.
+
+.. _selecting-the-active-environment:
+
+Selecting the Active Environment
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Symfony applications come with a file called ``.env`` located at the project
+root directory. This file is used to define the value of environment variables
+and it's explained in detail :ref:`later in this article <config-dot-env>`.
+
+Open the ``.env`` file and edit the value of the ``APP_ENV`` variable to change
+the environment in which the application runs. For example, to run the
+application in production:
+
+```bash
+# .env
+APP_ENV=prod
+```
+
+This value is used both for the web and for the console commands. However, you
+can override it for commands by setting the ``APP_ENV`` value before running them:
+
+.. code-block:: terminal
+
+    # Use the environment defined in the .env file
+    $ php bin/console command_name
+
+    # Ignore the .env file and run this command in production
+    $ APP_ENV=prod php bin/console command_name
+
+.. deprecated:: 4.2
+
+    In previous Symfony versions you could configure the environment with the
+    ``--env`` command option, which was deprecated in Symfony 4.2.
+
+Creating a New Environment
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The default three environments provided by Symfony are enough for most projects,
+but you can define more environments if you want to. For example, this is how
+you can define a ``staging`` environment where the client can test the project
+before going to production:
 
-For more information about parameters - including how to reference them from inside
-a controller - see :ref:`service-container-parameters`.
+#. Create a configuration directory with the same name as the environment (in
+   this case, ``config/packages/staging/``);
+#. Add the needed configuration files in ``config/packages/staging/`` to
+   define the behavior of the new environment. Symfony loads first the files in
+   ``config/packages/*.yaml``, so you must only configure the differences with
+   those files;
+#. Select the ``staging`` environment using the ``APP_ENV`` env var as explained
+   in the previous section.
 
+.. _config-env-vars:
+
+Configuration Based on Environment Variables
+--------------------------------------------
+
+Using `environment variables`_ (or "env vars" for short) is a common practice to
+configure sensitive options such as credentials and passwords. Instead of
+defining the values of those options in the configuration files, you add a
+reference to an environment variable which will be resolved at runtime (only
+once per request, to not impact performance).
```

Comment:
I'm against putting this recommendation (or even hinting at doing stuff like this) in the documentation. Using env vars to store sensitive options is a *bad* idea (see i.e. https://www.webfactory.de/blog/storing-secrets-for-symfony-applications ).

Instead: Env vars shine for settings that change per environment, e.g. a DATABASE_URL (which isn't sensitive).

We can then add a section explaining some of the env var processors (i.e. `file:` or `json:`) to show a better practice to store sensitive data.